### DISCUSSION TITLE
EntityRuler not working as expected

### DISCUSSION THREAD

**Discussion by albertovilla** [explosion/spaCy#6754] [Help: Coding & Implementations] [answered]
Hi,

I have a couple of doubts regarding `EntityRuler` and Spacy tokenization.

**Goal**: I'm trying to create entities to recognize part numbers expressions such as `PN: 1329NFN-899-12`

What I have been done so far is to create patterns to recognize PartNumberStructure as follows:

```python
patterns_pn_st = [
    { 'label': 'PartNumberStructure',
      'pattern': [
          { 'LOWER': 'pn' },
          { 'TEXT': ':', 'OP': '?'}
      ]
    }
]
```

Then I have created another pattern to identify the part numbers, basically whenever I have a PartNumberStructure entity followed by a string which follows a specific regex then that's a part number.

```python
patterns_pn = [
    { 'label': 'PartNumber',
      'pattern': [
          { 'ENT_TYPE': 'PartNumberStructure' },
          { 'TEXT': { 'REGEX': '[a-zA-Z0-9-_]+' } }
      ]
    }
]
```

The issues I have found are of two types. One of them is regarding the way Spacy makes the tokenization.

```python
doc = nlp("PN: 1329NFN-899-12")

for token in doc:
    print(token.text)
```

Result:

```
PN
:
1329NFN-899
-
12
```

The part number will not be identified as one unit as the tokenizer splits on the second dash. Why is the token split with the second and not the first? I don't know how to solve this issue because even if a use an operator to detect multiple instances of the regex i.e. `OP: +` that wouldn't work as the entity would recognize everything after "PN: " as a part number which si not what I want. Example PN: 1329NFN-899-12 is for a car engine would not be properly recognized.

The second issue I have is that the PartNumber pattern seems to not be working. With the above tokenization I would expect

```
PN: PartNumberStructure
1329NFN-899: PartNumber
```

But only the PartNumberStructure is detected but no PartNumber is detected. If I change the pattern removing the 'ENT_TYPE' part then all the tokens after PN: are detected.

The full code to test this behaviour is as follows:

```python
from spacy.lang.en import English
from spacy.pipeline import EntityRuler
from spacy import displacy

nlp = English()
ruler = EntityRuler(nlp)

patterns_pn_st = [
    { 'label': 'PartNumberStructure',
      'pattern': [
          { 'LOWER': 'pn' },
          { 'TEXT': ':', 'OP': '?'}
      ]
    }
]

patterns_pn = [
    { 'label': 'PartNumber',
      'pattern': [
          { 'ENT_TYPE': 'PartNumberStructure' },
          { 'TEXT': { 'REGEX': '[a-zA-Z0-9-_]+' } }
      ]
    }
]

# PartNumberForm patterns
ruler.add_patterns(patterns_pn_st)

# PartNumber patterns
ruler.add_patterns(patterns_pn)

# Add ruler to the pipeline
nlp.add_pipe(ruler)

doc = nlp("PN: 1329NFN-899-12")

for token in doc:
    print(token.text)
    
displacy.render(doc, style="ent", page=True)
```

This results in:

```
PN: PARTNUMBERSTRUCTURE 1329NFN-899-12
```

**Comment by albertovilla** [comment]
Hi, in order to solve the first of the issues, i.e. to avoid the tokenizer to split the part numbers with the hyphens I have tried to modified the infix rules of the tokenizer as follows:

```python
from spacy.lang.char_classes import LIST_ELLIPSES, LIST_ICONS, ALPHA_LOWER, ALPHA_UPPER, CONCAT_QUOTES, ALPHA, HYPHENS
from spacy.util import compile_infix_regex

_infixes = (
    LIST_ELLIPSES
    + LIST_ICONS
    + [
        r"(?<=[0-9])[+\*^](?=[0-9-])",
        r"(?<=[{al}{q}])\.(?=[{au}{q}])".format(
            al=ALPHA_LOWER, au=ALPHA_UPPER, q=CONCAT_QUOTES
        ),
        r"(?<=[{a}]),(?=[{a}])".format(a=ALPHA),
        r"(?<=[{a}0-9])[:<>=/](?=[{a}])".format(a=ALPHA),
    ]
)

infixes_regex = compile_infix_regex(_infixes)
nlp.tokenizer.infix_finditer = infixes_regex.finditer
```
Basically I have taken the original `_infixes` defined in `spacy.lang.punctuation.py` and I have modified two lines:
- the first line `r"(?<=[0-9])[+\-\*^](?=[0-9-])",` removing `\-`
- removing the line which covers the hyphens rules `r"(?<=[{a}])(?:{h})(?=[{a}])".format(a=ALPHA, h=HYPHENS),`

Then compile that list of regex using the function in `spacy.util` and then add the new function to `tokenizer.infix_finditer`.

WIth the modified tokenizer I have tried it with a part number:

```python
doc = nlp("PN: 1329NFN-899-12")

for token in doc:
    print(token.text)
```
And it works as expected:

```
PN
:
1329NFN-899-12
```

I found quite helpful to debug the function `explain` e.g. `nlp.tokenizer.explain("PN: 1329NFN-899-12")`

Maybe there is a simpler way to do it but this seems to be working fine. 

I still don't know how to solve the other issue, it must be easy as the only thing I want is to label any token after an specific entity type.

**Comment by albertovilla** [comment]
I think I have found a workaround to what I wanted to do, I'd rather implement it using `ENT_TYPE` but at least this works as I need. 

I have used a function which checks the current entities and whenever it finds a `PartNumberStructure` adds a new entity for `PartNumber` with the next token. I have added that to the pipeline with `add_pipe` after the `EntityRuler` and this solves the issue as it basically expands current entities with the part number ones as expected.

I believe the other approach is more flexible but as I cannot make it work I will use this one instead (at least for the moment).

**Reply by mehmetilker** [reply]
Have you tried "Matching regular expressions on the full text" if part number is in constant pattern or at least few different shape...-
https://spacy.io/usage/rule-based-matching#regex-text

**Comment by adrianeboyd** [comment, marked as answer]
The `EntityRuler` finds all the matches based on the initial `doc` state/annotation, so it doesn't work to have one rule depend on the label added by another rule.. Your custom component sounds like a good way to handle this, especially if you want the `PN:` and the number to be in two different entity spans.

**Reply by albertovilla** [reply]
Understood, I will use then the custom function in the pipeline, it works just fine, I just wanted to do it the "right" way from Spacy point of view and I was assuming that rule dependency was somewhat possible / recommended.
